File: src/numerical/potential/test_potential2D.m
function test_potential2D(filename)
narginchk(1,1)

addpath([fileparts(mfilename('fullpath')), '/../../../matlab'])

exist_or_skip(filename, 'file')

if isoctave
h = load(filename);

x2 = h.x2;
x3 = h.x3;
Phi = h.Phi;
Phi2 = h.Phi2squeeze;
Phitrue = h.Phitrue;
else
x2 = h5read('/x2');
x3 = h5read('/x3');
Phi = h5read('/Phi');
Phi2 = h5read('/Phi2squeeze');
Phitrue = h5read('/Phitrue');
end


assert_allclose(Phi2(13, 13), 0.00032659, 1e-5,[],'Potential 2d accuracy')

if ~isinteractive
  return
end

%% Plot data
figure(1);

subplot(1,3,1)
imagesc(x2,x3,Phi);
colorbar;
axis xy;
xlabel('distance (m)')
ylabel('distance (m)')
title('2D potential (polarization)')

subplot(1,3,2)
imagesc(x2,x3,Phi2);
colorbar;
axis xy;
xlabel('distance (m)')
ylabel('distance (m)')
title('2D potential (static)')

subplot(1,3,3)
imagesc(x2,x3,Phitrue);
colorbar;
axis xy;
xlabel('distance (m)')
ylabel('distance (m)')
title('2D potential (analytical)')

end

% fid=fopen(filename);
% data=fscanf(fid,'%f',1);
% lx2=data(1);
% x2=fscanf(fid,'%f',lx2);
% data=fscanf(fid,'%f',1);
% lx3=data(1);
% x3=fscanf(fid,'%f',lx3);
% Phi=fscanf(fid,'%f',lx2*lx3);
% Phi=reshape(Phi,[lx2,lx3]);
% Phi2=fscanf(fid,'%f',lx2*lx3);
% Phi2=reshape(Phi2,[lx2,lx3]);
% Phitrue=fscanf(fid,'%f',lx2*lx3);
% Phitrue=reshape(Phitrue,[lx2,lx3]);
% fclose(fid);

File: src/numerical/potential/test_potential2d.py
#!/usr/bin/env python3
import h5py
from pathlib import Path
import argparse
import sys
import numpy as np


def read_potential2D(fn: Path, doplot: bool = False):

    fn = Path(fn).expanduser()
    if not fn.is_file():
        print(fn, "not found", file=sys.stderr)
        raise SystemExit(77)

    with h5py.File(fn, "r") as f:
        lx1 = f["/lx1"][()]
        lx2 = f["/lx2"][()]
        lx3 = f["/lx3"][()]
        x1 = f["/x1"][:]
        x2 = f["/x2"][:]
        x3 = f["/x3"][:]
        Phi = f["/Phi"][:]
        Phi2 = f["/Phi2squeeze"][:]
        Phitrue = f["/Phitrue"][:]
    assert np.isclose(Phi2[12, 12], 0.00032659, 1e-3), "Potential 2d accuracy"
    assert lx1 == x1.size
    assert lx2 == x2.size
    assert lx3 == x3.size

    if not doplot:
        return

    fg = figure(figsize=(15, 6))
    ax = fg.subplots(1, 3, sharey=True)
    h = ax[0].pcolormesh(x2, x3, Phi)
    fg.colorbar(h, ax=ax[0])
    ax[0].set_ylabel("distance [m]")
    ax[0].set_xlabel("distance [m]")
    ax[0].set_title("2D potential (polarization)")

    h = ax[1].pcolormesh(x2, x3, Phi2)
    fg.colorbar(h, ax=ax[1])
    ax[1].set_title("2D potential (static)")

    h = ax[2].pcolormesh(x2, x3, Phitrue)
    fg.colorbar(h, ax=ax[2])
    ax[2].set_title("2D potential (analytical)")

    # with fn.open("r") as f:
    #     (lx2,) = np.fromfile(f, int, 1, sep=" ")
    #     x2 = np.fromfile(f, float, lx2, sep=" ")
    #     (lx3,) = np.fromfile(f, int, 1, sep=" ")
    #     x3 = np.fromfile(f, float, lx3, sep=" ")
    #     Phi = np.fromfile(f, float, lx2 * lx3, sep=" ").reshape((lx2, lx3))
    #     Phi2 = np.fromfile(f, float, lx2 * lx3, sep=" ").reshape((lx2, lx3))
    #     Phitrue = np.fromfile(f, float, lx2 * lx3, sep=" ").reshape((lx2, lx3))
    # assert np.isclose(Phi2[12, 12], 0.000327, 1e-5), "Potential 2d accuracy"


if __name__ == "__main__":
    p = argparse.ArgumentParser()
    p.add_argument("file")
    p.add_argument("-p", "--plot", help="make plots", action="store_true")
    P = p.parse_args()

    if P.plot:
        from matplotlib.pyplot import figure, show

    read_potential2D(P.file, P.plot)

    if P.plot:
        show()
